### PR TITLE
Version `0.4`

### LINKED ISSUE #111: [Event]: activespecial
### What is the name of the event

activespecial

### LINKED ISSUE #173: [Event]: configreloaded
### What is the name of the event

configreloaded

### LINKED ISSUE #174: [Event]: monitoraddedv2
### What is the name of the event

monitoraddedv2

### LINKED ISSUE #175: [Event]: ignoregrouplock
### What is the name of the event

ignoregrouplock

### LINKED ISSUE #176: [Event]: lockgroups
### What is the name of the event

lockgroups

### PR DESCRIPTION
**Pull request by yavko**
Done with:
- [x] All new events
- [x] All new hyprctl commands
- [ ] All new dispatchers
- [ ] All new info commands
- [x] Adding `Stream` API
- [x] Removing regexp matching
- [x] Making it really easy to add new events

### CODE REVIEW

**Comment by yavko** [comment]
I'm going to need help with testing, since I'm not on hyprland

**Comment by luyu-wu** [comment]
activespecial lgtm. I'll try and add workspacev2 later since that's important to what I'm working on rn!

**Comment by yavko** [comment]
> activespecial lgtm. I'll try and add workspacev2 later since that's important to what I'm working on rn!

Thanks! Someone else is also working on configreloaded, so I'll probably have to work on the rest

**Comment by cyrinux** [comment]
@yavko maybe it would be better to have seperate PR for each event ?

**Comment by yavko** [comment]
> @yavko maybe it would be better to have seperate PR for each event ?

Sure, ig, if anyone else wants to add the events they can, all the ones I do will be in this PR

**Comment by yavko** [comment]
I will merge this once I finish adding all the left over dispatcher and info commands, and i get 2 approving reviews and tests. I am not able to test this PR, as I'm not running hyprland.

**Comment by yavko** [comment]
Since most of the changes are with the event system, i would like testing more on that, and on the new `EventStream` api.

File: src/shared.rs [hunk 1]
```diff
@@ -109,7 +105,12 @@ impl Address {
     }
     /// This creates a new address from a value that implements [std::string::ToString]
     pub fn new<T: ToString>(string: T) -> Self {
-        Self(string.to_string())
+        let str = string.to_string();
+        if str.starts_with("0x") {
+            Self(str)
+        } else {
+            Self("0x".to_owned() + str.as_str())
```
Comment (diniamo) [inline src/shared.rs:112]: Use format!

Comment (yavko) [reply src/shared.rs:112]: Format is actually slower in this case, the original code way back when, used `format!`

Comment (JakeStanger) [reply src/shared.rs:112]: would probs be wise to stick a comment on it saying that so nobody tries to refactor it in the future

File: src/event_listener/shared.rs [hunk 2]
```diff
@@ -502,364 +614,309 @@ fn parse_string_as_work(str: String) -> WorkspaceType {
...
+    "activelayout" => ((2),ParsedEventType::LayoutChanged),
+    "activespecial" => ((2),ParsedEventType::ActiveSpecial),
+    "submap" => ((1), ParsedEventType::SubMapChanged),
+    "openlayer" => ((1),ParsedEventType::LayerOpened),
+    "closelayer" => ((1),ParsedEventType::LayerClosed),
+    "changefloatingmode" => ((2),ParsedEventType::FloatStateChanged),
+    "minimize" => ((2),ParsedEventType::Minimize),
+    "screencast" => ((2),ParsedEventType::Screencast),
+    "urgent" => ((1),ParsedEventType::UrgentStateChanged),
+    "windowtitlev2" => ((1),ParsedEventType::WindowTitleChangedV2),
```
Comment (NBonaparte) [inline src/event_listener/shared.rs:680]: This should be 2 (https://wiki.hyprland.org/IPC/#events-list), fixes the error reported in https://github.com/hyprland-community/hyprland-rs/pull/242
```suggestion
    "windowtitlev2" => ((2),ParsedEventType::WindowTitleChangedV2),
```

File: src/event_listener/shared.rs [hunk 3]
```diff
@@ -502,364 +614,309 @@ fn parse_string_as_work(str: String) -> WorkspaceType {
...
+    "renameworkspace" => ((2),ParsedEventType::WorkspaceRename),
+    "focusedmon" => ((2),ParsedEventType::ActiveMonitorChanged),
+    "activewindow" => ((2),ParsedEventType::ActiveWindowChangedV1),
+    "activewindowv2" => ((1),ParsedEventType::ActiveWindowChangedV2),
+    "fullscreen" => ((1),ParsedEventType::FullscreenStateChanged),
+    "monitorremoved" => ((1),ParsedEventType::MonitorRemoved),
+    "monitoraddedv2" => ((3),ParsedEventType::MonitorAddedV2),
+    "openwindow" => ((4),ParsedEventType::WindowOpened),
+    "closewindow" => ((1),ParsedEventType::WindowClosed),
+    "movewindow" => ((3),ParsedEventType::WindowMovedV2),
```
Comment (NBonaparte) [inline src/event_listener/shared.rs:670]: This should also be `movewindowv2`:
```suggestion
    "movewindowv2" => ((3),ParsedEventType::WindowMovedV2),
```

File: src/event_listener/shared.rs [hunk 4]
```diff
@@ -502,364 +614,309 @@ fn parse_string_as_work(str: String) -> WorkspaceType {
...
+    "activelayout" => ((2),ParsedEventType::LayoutChanged),
+    "activespecial" => ((2),ParsedEventType::ActiveSpecial),
+    "submap" => ((1), ParsedEventType::SubMapChanged),
+    "openlayer" => ((1),ParsedEventType::LayerOpened),
+    "closelayer" => ((1),ParsedEventType::LayerClosed),
+    "changefloatingmode" => ((2),ParsedEventType::FloatStateChanged),
+    "minimize" => ((2),ParsedEventType::Minimize),
```
Comment (NBonaparte) [inline src/event_listener/shared.rs:677]: I don't see minimized anywhere in the table.

File: src/event_listener/shared.rs (hunk 2 above)
Comment (yavko) [reply src/event_listener/shared.rs:680]: Thx

File: src/event_listener/shared.rs (hunk 4 above)
Comment (yavko) [reply src/event_listener/shared.rs:677]: I'm honestly not sure why I added this, maybe it was removed, I mean it's not even in the source code. https://github.com/hyprland-community/hyprland-rs/commit/f177a52fba3d6f0848b7543f1dd510ca9e14fc9e

Comment (yavko) [reply src/event_listener/shared.rs:677]: The event needs to get removed lol

**Comment by yavko** [comment]
hopefully everything works now

**Comment by NBonaparte** [comment]
I've been running the listener for a couple days while using Hyprland normally and haven't seen any other issues yet.

**Comment by yavko** [comment]
> I've been running the listener for a couple days while using Hyprland normally and haven't seen any other issues yet.

Sick imma get ready to merge then

**Comment by yavko** [comment]
I've decided to put the rest of the features in a different PR, and so i am actually just going to merge this now

**Pull request merged**